triedb/pathdb/history_indexer.go

```diff
@@ -0,0 +1,529 @@
+// Copyright 2025 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/
+
+package pathdb
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// The batch size for reading state histories
+const historyReadBatch = 1000
+
+// batchIndexer is a structure designed to perform batch indexing or unindexing
+// of state histories atomically.
+type batchIndexer struct {
+	accounts map[common.Address][]uint64                 // History ID list, Keyed by account address
+	storages map[common.Address]map[common.Hash][]uint64 // History ID list, Keyed by account address and the hash of raw storage key
+	counter  int                                         // The counter of processed states
+	delete   bool                                        // Index or unindex mode
+	lastID   uint64                                      // The ID of latest processed history
+	db       ethdb.KeyValueStore
+}
+
+// newBatchIndexer constructs the batch indexer with the supplied mode.
+func newBatchIndexer(db ethdb.KeyValueStore, delete bool) *batchIndexer {
+	return &batchIndexer{
+		accounts: make(map[common.Address][]uint64),
+		storages: make(map[common.Address]map[common.Hash][]uint64),
+		delete:   delete,
+		db:       db,
+	}
+}
+
+// process iterates through the accounts and their associated storage slots in the
+// state history, tracking the mapping between state and history IDs.
+func (b *batchIndexer) process(h *history, historyID uint64) error {
+	for _, address := range h.accountList {
+		b.counter += 1
+		b.accounts[address] = append(b.accounts[address], historyID)
```

Comment:
Just thinking out loud, the whole following code depends on the historyIDs to be sorted. So if we ever called process with an out of order historyID, this would not work anymore. Should we make sure before writing out the lists, that they are sorted?

Reply:
The histories are resolved from the freezer in batch. I think the assumption is held that histories are processed in order.